gui/guiComponent.py

```diff
@@ -0,0 +1,225 @@
+import os
+from tkinter import *
+import screeninfo
+import time
+from utils.config import USE_GUI, TIMEOUT_GUI
+
+# We do not need this on Linux.
+if os.name == "nt":
+    import pythoncom
+    import win32com.client
+    import win32gui
+
+
+def windowEnumerationHandler(hwnd, top_windows):
+    """
+    Handler for Windows OS enumeration of all open windows.
+    Used to return to the Path of Exile window after displaying the overlay.
+    """
+    top_windows.append((hwnd, win32gui.GetWindowText(hwnd)))
+
+
+def windowToFront(root):
+    # This is necessary for displaying the GUI window above active window(s) on the Windows OS
+    if os.name == "nt":
+        # In order to prevent SetForegroundWindow from erroring, we must satisfy the requirements
+        # listed here:
+        # https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setforegroundwindow
+        # We satisfy this by internally sending the alt character so that Windows believes we are
+        # an active window.
+        # We need this pythoncom call for win32com use in a thread.
+        pythoncom.CoInitialize()
+        shell = win32com.client.Dispatch("WScript.Shell")
+        shell.SendKeys("%")
+        win32gui.SetForegroundWindow(root.winfo_id())
+
+
+def windowRefocus(name):
+    """
+    Restore focus to a window, if on Windows.
+    TODO: If originating window was NOT "name", return to previous window.
+    """
+
+    if os.name == "nt":
+        results = []
+        top_windows = []
+        win32gui.EnumWindows(windowEnumerationHandler, top_windows)
+        for i in top_windows:
+            if name == i[1].lower():
+                win32gui.ShowWindow(i[0], 5)
+                win32gui.SetForegroundWindow(i[0])
+                break
+
+components = []
+
+class GuiComponent:
+    def __init__(self):
+        self.frame = None
+        self.closed = True
+        self.opened = time.time()
+        self.elapsed = 0
+        self.have_timeout = False
+        self.hidden = True
+        components.append(self)
+
+    def stop(self):
+        self.frame.quit()
+        self.close()
+
+    def run(self):
+        self.frame.mainloop()
+
+    def should_close(self):
+        if self.is_closed():
+            return
+        if not self.have_timeout:
+            return
+        self.elapsed = time.time() - self.opened
+        if self.elapsed >= int(TIMEOUT_GUI):
+            elapsed = 0
+            self.close()
+            windowRefocus("path of exile")
+
+    def should_hide(self):
+        if not self.have_timeout:
+            return
+        self.elapsed = time.time() - self.opened
+        if self.elapsed >= int(TIMEOUT_GUI):
+            elapsed = 0
+            self.hide()
+    
+    def prepare_window(self):
+        tk = Tk().withdraw()
```

Comment:
Is this really necessary? As this looks like you are making empty hidden windows every time prepare_window is called

Reply:
Calling Tk() every time also really eats up memory on Linux machines. I learned that the hard way at the start of this project when I was getting 1000's of tk windows popping up.